win32ss/user/ntuser/ghost.c

```diff
@@ -62,89 +119,280 @@ BOOL FASTCALL IntIsGhostWindow(PWND Window)
+    Cs.y = rc.top;
+    Cs.x = rc.left;
+    Cs.style = style;
+    Cs.lpszName = NULL;
+    Cs.lpszClass = GhostClass.Buffer;
+    Cs.dwExStyle = exstyle;
+
+    // WindowName
+    RtlZeroMemory(&WindowName, sizeof(WindowName));
+
+    // create the ghost
+    pGhostWnd = co_UserCreateWindowEx(&Cs, &GhostClass, &WindowName, NULL);
+    if (!pGhostWnd)
+    {
+        ERR("Failed to create ghost for %p\n", (void *)hwndTarget);
+        goto Quit;
+    }
+
+    // add to list
+    pEntry = IntAddGhostEntry(hwndTarget, pGhostWnd->head.h);
+
+Quit:
+    if (!pEntry && pGhostWnd)
+    {
+        // destroy ghost
+        co_UserDestroyWindow(pGhostWnd);
+    }
+
+    // resume hooks
+    if (!NoHooks)
+    {
+        ptiCurrent->TIF_flags &= ~TIF_DISABLEHOOKS;
+        ptiCurrent->pClientInfo->dwTIFlags = ptiCurrent->TIF_flags;
+    }
+
+    return pEntry;
 }
 
-HWND FASTCALL UserGhostWindowFromHungWindow(HWND hwndHung)
+BOOL FASTCALL
+IntGhostOnDestroy(HWND hwndGhost, HWND hwndTarget)
 {
-    PWND pHungWnd = ValidateHwndNoErr(hwndHung);
-    if (!pHungWnd)
+    PWND pGhostWnd;
+
+    pGhostWnd = ValidateHwndNoErr(hwndGhost);
+    if (!pGhostWnd)
     {
-        DPRINT("Not a window\n");
-        return NULL;
+        ERR("Not a window (%p, %p).\n", (void *)hwndTarget, (void *)hwndGhost);
+        return FALSE;
     }
-    return IntGhostWindowFromHungWindow(pHungWnd);
+
+    if (!IntIsGhostWindow(pGhostWnd))
+    {
+        ERR("Not a Ghost window (%p, %p)\n", (void *)hwndTarget, (void *)hwndGhost);
+        return FALSE;
+    }
+
+    return IntDeleteGhostEntry(hwndTarget);
 }
 
-HWND FASTCALL IntHungWindowFromGhostWindow(PWND pGhostWnd)
+VOID FASTCALL IntFreeGhostList(VOID)
 {
-    const GHOST_DATA *UserData;
-    HWND hwndTarget;
+    PLIST_ENTRY CurrentEntry, NextEntry, ListHead = &gGhostListHead;
+    PGHOST_ENTRY pEntry;
 
-    if (!IntIsGhostWindow(pGhostWnd))
+    for (CurrentEntry = ListHead->Flink;
+         CurrentEntry != ListHead;
+         CurrentEntry = NextEntry)
     {
-        DPRINT("Not a ghost window\n");
-        return NULL;
+        pEntry = CONTAINING_RECORD(CurrentEntry, GHOST_ENTRY, ListEntry);
+        NextEntry = CurrentEntry->Flink;
+        ExFreePoolWithTag(pEntry, USERTAG_GHOST);
     }
 
-    UserData = (const GHOST_DATA *)pGhostWnd->dwUserData;
-    if (UserData)
+    InitializeListHead(&gGhostListHead);
+}
+
+/*
+ * GhostThreadMain
+ *
+ * Creates ghost windows and exits when no non-responsive window remains.
+ */
+VOID NTAPI
+UserGhostThreadEntry(VOID)
+{
+    // This thread should handle all ghost windows and exit when no ghost window is needed.
+    MSG msg;
+    HDESK hDesk;
+
+    TRACE("UserGhostThreadEntry: started.\n");
+
+    ASSERT(!UserIsEntered());
+    UserEnterExclusive();
+
+    // attach thread to desktop
+    hDesk = IntGetDesktopObjectHandle(gpdeskInputDesktop);
+    IntSetThreadDesktop(hDesk, FALSE);
+
+    // get a PTHREADINFO
+    FIXME("We're missing a thread lock\n");
```

Comment:
A comment would probably be enough, sorry for not being clear.

Reply:
OK.